File: LinkList/25.reverse-nodes-in-k-group.py
from typing import List, Optional
from collections import defaultdict
from type import ListNode


# @leet start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, next=head)
        preGroup = dummy

        while True:
            kth = self.getNth(preGroup, k)
            if not kth:
                break

            nextGroup = kth.next

            prev, cur = nextGroup, preGroup.next
            while (
                cur != nextGroup
            ):  # NOTE: not cur != kth.next. kth will be update later
                temp = cur.next
                cur.next = prev
                prev = cur
                cur = temp

            temp = preGroup.next
            preGroup.next = kth
            preGroup = temp

        return dummy.next

    def getNth(self, cur, k):
        while cur and k > 0:
            cur = cur.next
            k -= 1
        return cur


# @leet end

File: LinkList/25.reverse-nodes-in-k-group.go
package leetcode

// @leet start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
	var (
		dummy  = &ListNode{Next: head}
		preGrp = dummy
		getNth = func(cur *ListNode, k int) *ListNode {
			for cur != nil && k > 0 {
				cur = cur.Next
				k--
			}
			return cur
		}
	)
	for {
		kth := getNth(preGrp, k)
		if kth == nil {
			break
		}
		nextGrp := kth.Next
		prev, cur := nextGrp, preGrp.Next
		for cur != nextGrp {
			temp := cur.Next
			cur.Next = prev
			prev = cur
			cur = temp
		}
		preGrp, preGrp.Next = preGrp.Next, kth
		// temp := preGrp.Next
		// preGrp.Next = kth
		// preGrp = temp
	}
	return dummy.Next
}

// @leet end

